File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/rsp/RspAccountList.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in the root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.rsp

import com.mytiki.sdk.capture.receipt.capacitor.Account
import org.json.JSONArray
import org.json.JSONObject

/**
 * Represents the response containing a list of [Account] objects.
 *
 * @param accounts The list of [Account] objects.
 * @param error An optional [Exception] object representing an error, if any.
 */
class RspAccountList(
    private val accounts: MutableList<Account>,
    private val error: Exception? = null
) : Rsp {

    /**
     * Converts this response to a JSON object.
     *
     * @return A [JSONObject] representing this response.
     */
    override fun toJson(): JSONObject =
        JSONObject()
            .put("accounts", JSONArray(accounts.map { account -> account.toRsp() }))
            .put("error", error?.message ?: "")
}

File: ios/Plugin/Rsp/RspAccountList.swift
/**
 Represents a response containing a list of Account objects for the ReceiptCapture plugin.
 
 This struct is used to encapsulate a list of accounts and an optional error message in response to plugin calls that retrieve account information.
 */
struct RspAccountList: Rsp {
    /// An array of `Account` objects representing the list of accounts.
    let accounts: [Account]
    
    /// An optional error message, which is `nil` when no error occurs.
    let error: String? = nil

    /**
     Converts the `RspAccountList` object into a dictionary suitable for use in plugin response data.
     
     - Returns: A dictionary representing the list of accounts and an optional error message in a format suitable for a Capacitor plugin call result.
     */
    func toPluginCallResultData() -> [String: Any] {
        var result: [String: Any] = [:]
        result["accounts"] = accounts.map { account in account.toResultData() }
        result["error"] = error
        return result
    }
}
